unsupported_operation_exception on collapse + docvalue_fields

### Elasticsearch Version

8.8.0

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux 5.19.0-1022-aws #23~22.04.1-Ubuntu SMP Fri Mar 17 15:38:24 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Problem Description

ES returns unsupported_operation_exception: null instead of a descriptive error message for a minimal query with collapse and docvalue_fields

```
{
  "took" : 196,
  "timed_out" : false,
  "_shards" : {
    "total" : 64,
    "successful" : 63,
    "skipped" : 0,
    "failed" : 1,
    "failures" : [
      {
        "shard" : 0,
        "index" : "idx",
        "node" : "rcZ4pURhRNSeSL349CCG-Q",
        "reason" : {
          "type" : "unsupported_operation_exception",
          "reason" : "unsupported_operation_exception: null"
        }
      }
    ]
  },
  "hits" : {
    "total" : {
      "value" : 10000,
      "relation" : "gte"
    },
    "max_score" : null,
    "hits" : [ ]
  }
}
```

### Steps to Reproduce

```
{
 "collapse": {"field": "col1"},
  "docvalue_fields": [
  "col2",
  "col3"
 ]
}
```

### Logs (if relevant)

_No response_

Any news regarding this issue?
Still there in 8.12.2, seems impossible to use docvalue_fields when using collapse without getting this error:

Caused by: java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractCollection.add(AbstractCollection.java:253)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.fetch.FetchContext.docValuesContext(FetchContext.java:179)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.fetch.subphase.FetchDocValuesPhase.getProcessor(FetchDocValuesPhase.java:33)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.fetch.FetchPhase.getProcessors(FetchPhase.java:183)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:108)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:78)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:717)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:688)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:545)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:73)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:70)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:95)
	at org.elasticsearch.server@8.12.2/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more

--- a/docs/changelog/110103.yaml
+++ b/docs/changelog/110103.yaml
@@ -2,4 +2,4 @@ pr: 110103
 summary: Fix automatic tracking of collapse with `docvalue_fields`
 area: Search
 type: bug
-issues: []
+issues: [96510]

--- a/docs/changelog/110103.yaml
+++ b/docs/changelog/110103.yaml
@@ -2,4 +2,4 @@ pr: 110103
 summary: Fix automatic tracking of collapse with `docvalue_fields`
 area: Search
 type: bug
-issues: [96510]
+issues: []

--- a/docs/changelog/110103.yaml
+++ b/docs/changelog/110103.yaml
@@ -1,0 +1,5 @@
+pr: 110103
+summary: Fix automatic tracking of collapse with `docvalue_fields`
+area: Search
+type: bug
+issues: []

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/CollapseSearchResultsIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/CollapseSearchResultsIT.java
@@ -39,4 +39,26 @@ public class CollapseSearchResultsIT extends ESIntegTestCase {
             }
         );
     }
+
+    public void testCollapseWithDocValueFields() {
+        final String indexName = "test_collapse";
+        createIndex(indexName);
+        final String collapseField = "collapse_field";
+        final String otherField = "other_field";
+        assertAcked(indicesAdmin().preparePutMapping(indexName).setSource(collapseField, "type=keyword", otherField, "type=keyword"));
+        index(indexName, "id_1_0", Map.of(collapseField, "value1", otherField, "other_value1"));
+        index(indexName, "id_1_1", Map.of(collapseField, "value1", otherField, "other_value2"));
+        index(indexName, "id_2_0", Map.of(collapseField, "value2", otherField, "other_value3"));
+        refresh(indexName);
+
+        assertNoFailuresAndResponse(
+            prepareSearch(indexName).setQuery(new MatchAllQueryBuilder())
+                .addDocValueField(otherField)
+                .setCollapse(new CollapseBuilder(collapseField).setInnerHits(new InnerHitBuilder("ih").setSize(2))),
+            searchResponse -> {
+                assertEquals(collapseField, searchResponse.getHits().getCollapseField());
+                assertEquals(Set.of(new BytesRef("value1"), new BytesRef("value2")), Set.of(searchResponse.getHits().getCollapseValues()));
+            }
+        );
+    }
 }

--- a/server/src/main/java/org/elasticsearch/search/fetch/FetchContext.java
+++ b/server/src/main/java/org/elasticsearch/search/fetch/FetchContext.java
@@ -189,7 +189,7 @@ public class FetchContext {
                     searchContext.getSearchExecutionContext(),
                     Collections.singletonList(new FieldAndFormat(name, null))
                 );
-            } else if (searchContext.docValuesContext().fields().stream().map(ff -> ff.field).anyMatch(name::equals) == false) {
+            } else if (searchContext.docValuesContext().fields().stream().map(ff -> ff.field).noneMatch(name::equals)) {
                 dvContext.fields().add(new FieldAndFormat(name, null));
             }
         }

--- a/server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java
+++ b/server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java
@@ -10,6 +10,7 @@ package org.elasticsearch.search.fetch.subphase;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.query.SearchExecutionContext;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -24,7 +25,7 @@ import java.util.Map;
  */
 public class FetchDocValuesContext {
 
-    private final Collection<FieldAndFormat> fields;
+    private final List<FieldAndFormat> fields;
 
     /**
      * Create a new FetchDocValuesContext using the provided input list.
@@ -40,7 +41,7 @@ public class FetchDocValuesContext {
                 fieldToFormats.put(fieldName, new FieldAndFormat(fieldName, field.format, field.includeUnmapped));
             }
         }
-        this.fields = fieldToFormats.values();
+        this.fields = new ArrayList<>(fieldToFormats.values());
         int maxAllowedDocvalueFields = searchExecutionContext.getIndexSettings().getMaxDocvalueFields();
         if (fields.size() > maxAllowedDocvalueFields) {
             throw new IllegalArgumentException(
@@ -58,7 +59,7 @@ public class FetchDocValuesContext {
     /**
      * Returns the required docvalue fields.
      */
-    public Collection<FieldAndFormat> fields() {
+    public List<FieldAndFormat> fields() {
         return this.fields;
     }
 }

